### PR TITLE
[stdlib] Implement native grapheme breaking for String

### PR DESCRIPTION
**Pull request by Azoy**
This is a draft and will see many updates, but wanted to put this up to get some benchmark readings.

Resolves: rdar://52194063, https://bugs.swift.org/browse/SR-9423

### CODE REVIEW

**Comment by Azoy** [comment]
@swift-ci Please benchmark

**Comment by Azoy** [comment]
@swift-ci Please test

**Comment by Azoy** [comment]
@swift-ci Please benchmark

**Comment by Azoy** [comment]
@swift-ci Please benchmark
@swift-ci Please test

**Comment by Azoy** [comment]
@swift-ci Please benchmark

**Comment by Azoy** [comment]
@swift-ci Please benchmark

**Comment by Azoy** [comment]
@swift-ci Please benchmark

**Comment by Azoy** [comment]
@swift-ci please benchmark

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:41]: Tiny code smell: Why is this cached and is it because we're trying to fit into an interface that doesn't quite map to what we're doing?

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:39]: Tiny code smell: the fact that both are stored, vary across invocations, and are convertible to each other

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:44]: Smell: probably clearer to make a `State` struct and parameterize over it. What state is relevant for breaking vs walking the scalars?

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:49]: Should we assert that the index is scalar aligned?

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:71]: Should we be asserting at all here? That is, these vary across calls to methods, but they probably have invariants in how they're expected to vary.

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:84]: Super early feedback: where do the fast paths live? This could be a function taking scalars that checks fast-paths and then falls back to getting the properties.

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:101]: Is it possible to run off the front of the string?

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:95]: Might this run off the front of the string?

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:75]: I wonder if it's helpful to have a helper method that does both. E.g.

```
while let sc1 = advanceScalar() { ...
```

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:77]: Couldn't `index` be at the end here?

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:110]: Definitely seems worthwhile to extract into a helper method

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:132]: Yeah, it might be cleaner to wrap up the state in a separate struct. There's a lot that can be done to an `inout` iterator beyond setting these bools (e.g. advancing the index)

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:176]: This feels like a complicated cross-call convention, so hopefully we can clarify it somewhat.

File: stdlib/public/core/UnicodeGraphemeBreakProperty.swift
Comment (milseman) [inline stdlib/public/core/UnicodeGraphemeBreakProperty.swift:16]: Let's underscore this. We will probably want to make API one day and that makes it easier.

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:35]:
```suggestion
internal struct _GraphemeIterator {
```

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:89]: What if this just trafficked in indices, and the caller can do this?

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:103]: Similarly, it might be useful for a helper. Hmm, maybe "retreatScalar"? naming is hard...

File: stdlib/public/core/GraphemeIterator.swift
Comment (milseman) [inline stdlib/public/core/GraphemeIterator.swift:128]: A common pattern is to have many forwards calls in a row or many backwards calls in a row. Does it make sense to have a notion of direction, or even different state structs for direction?

**Review by milseman — COMMENTED** [review]
This is a quick preliminary review, so many of my suggestions are vague and might not actually make sense. But, hopefully it gives you something to chew on or encourages some refactoring or clarifying. 

Looking good so far though!

File: stdlib/public/core/GraphemeIterator.swift
Comment (Azoy) [reply stdlib/public/core/GraphemeIterator.swift:101]: At the moment, no, because `_GraphemeIterator` has a set of assumptions from the caller. The only callers of this are `_opaqueCharacterStride` and they themselves check if there are at least 2 indices (including the given index and one forward/backward). I can try to rewrite this to be more general and not have these kinds of assumptions.

File: stdlib/public/core/GraphemeIterator.swift
Comment (Azoy) [reply stdlib/public/core/GraphemeIterator.swift:77]: Similar answer here, but the loop ensures that there's at least another index after the current one.

File: stdlib/public/core/GraphemeWalker.swift
Comment (milseman) [inline stdlib/public/core/GraphemeWalker.swift:87]:
```suggestion
  // Returns the index of the next grapheme cluster boundary
```

File: stdlib/public/core/GraphemeWalker.swift
Comment (milseman) [inline stdlib/public/core/GraphemeWalker.swift:107]: This looks **so** much clearer, thanks for the refactor!

File: stdlib/public/core/GraphemeWalker.swift
Comment (milseman) [inline stdlib/public/core/GraphemeWalker.swift:75]: Stick a tuple label on the return value to make it clear what that index is. Is it the start or end of the returned scalar? How is it different to an updated state?

File: stdlib/public/core/GraphemeWalker.swift
Comment (milseman) [inline stdlib/public/core/GraphemeWalker.swift:89]: What are we mutating?

File: stdlib/public/core/GraphemeWalker.swift
Comment (milseman) [inline stdlib/public/core/GraphemeWalker.swift:147]: Nice. I wonder if you want to outline the rest of this function too.

File: stdlib/public/core/GraphemeWalker.swift
Comment (milseman) [inline stdlib/public/core/GraphemeWalker.swift:288]: But is State updated here, or can it be copied in?

**Review by milseman — COMMENTED** [review]
Quick review. This is looking a lot cleaner, thanks for scrapping the iterator pattern!

Refactoring out state helps with the clarity quite a bit. It also shows that `index` does seem like a different bit of state than the others, so we might pull that out and put it back on the walker or somewhere. Something to consider.

File: stdlib/public/core/GraphemeWalker.swift
Comment (Azoy) [reply stdlib/public/core/GraphemeWalker.swift:288]: State's `isInEmojiSequence` gets updated here.

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
@swift-ci please benchmark

**Review by milseman — COMMENTED** [review]
Sorry it's hasty, but I wanted to get you some feedback quickly.

**Review by milseman — COMMENTED** [review]
This is very close!

For landing, I think we can simplify a lot of the state that's passed around and mutated, especially when going backwards.

For future work, I'm interested in splitting this between forwards and backwards a little (since I think that will actually simplify the logic), but I don't want to hold up landing this for that. Those comments are written with "Future note:" at the front.

File: stdlib/public/core/StringGraphemeBreaking.swift
Comment (Azoy) [reply stdlib/public/core/StringGraphemeBreaking.swift:320]: Emoji state is needed for both going forward and backwards. Going backwards, this emoji state lets us continuously walk further and further backwards to find the grapheme index before the emoji.

Comment (Azoy) [reply stdlib/public/core/StringGraphemeBreaking.swift:320]: One example of an emoji sequence that Unicode has is the following:

"A👶  🏿  ◌̈  □  👶  🏿"

==

[.any, .extendedPictographic, .extend, .extend, .zwj, .extendedPictographic, .extend]


If we start at the boundary at the last .extend, our first rule is `(.extP, .extend)` which evaluates to false, but also updates the state bit here. We keep walking until we see the A and the first emoji which is `(.any, .extP)` which is finally true.

I can change the condition to be: `if state.isBackwards, !state.isInEmojiSequence` because in this example we walk backwards for no reason because we've already evaluated that we're in an emoji sequence with the first iteration.

File: stdlib/public/core/StringGraphemeBreaking.swift
Comment (Azoy) [reply stdlib/public/core/StringGraphemeBreaking.swift:389]: I commented on this in a different reply.

**Comment by Azoy** [comment]
@swift-ci please test

**Comment by Azoy** [comment]
@swift-ci please benchmark

File: stdlib/public/core/StringGraphemeBreaking.swift
Comment (milseman) [inline stdlib/public/core/StringGraphemeBreaking.swift:353]: Maybe add a comment describing what we're doing and why

File: stdlib/public/core/StringGraphemeBreaking.swift
Comment (milseman) [inline stdlib/public/core/StringGraphemeBreaking.swift:362]: Maybe a comment, or a more descriptive name like checkPairedRIs.

**Review by milseman — APPROVED** [review]
LGTM. Just a couple places for comments. I think we should try to merge this soon. There's more future work we could do, but this is great for now!

**Comment by milseman** [comment]
@swift-ci please test linux platform

**Comment by xwu** [comment]
Are the `-Osize` regressions in the x86_64 `Suffix*` benchmarks a problem?

**Comment by Azoy** [comment]
I have no idea why those benchmarks are showing as a regression in this patch (and it seems they showed a 3x improvement with the normalization patch). Considering they don't even touch strings at all, I'm inclined to say they are not an issue, but I could be wrong.

**Comment by Azoy** [comment]
Here are some benchmarks that I ran locally with the `StringWalk` benchmarks enabled:

### Performance (arm64): -O

**Regression**                             | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                                       | ---:   | ---:  | ---:   | ---:     
ArrayAppendGenericStructs                  | 920    | 1220  | +32.6% | **0.75x (?)**
CharIteration_utf16_unicodeScalars         | 1720   | 2040  | +18.6% | **0.84x (?)**
ParseFloat.Float.Exp                       | 6      | 7     | +16.7% | **0.86x (?)**
Set.isSubset.Seq.Int.Empty                 | 82     | 93    | +13.4% | **0.88x (?)**
Set.isStrictSubset.Seq.Int.Empty           | 82     | 93    | +13.4% | **0.88x (?)**
ArrayPlusEqualArrayOfInt                   | 180    | 200   | +11.1% | **0.90x (?)**
ArrayAppendToGeneric                       | 190    | 210   | +10.5% | **0.90x (?)**
BufferFillFromSlice                        | 10     | 11    | +10.0% | **0.91x (?)**
StringWalk_punctuated_characters_Backwards | 440    | 480   | +9.1%  | **0.92x (?)**
StringWalk_ascii_characters_Backwards      | 1800   | 1960  | +8.9%  | **0.92x (?)**
DataCreateEmptyArray                       | 650    | 700   | +7.7%  | **0.93x (?)**
&nbsp; | | | | 
**Improvement**                            | **OLD** | **NEW** | **DELTA** | **RATIO**
SubstringTrimmingASCIIWhitespace           | 314    | 86    | -72.6% | **3.65x**
StringWalk_utf16_characters                | 26080  | 7200  | -72.4% | **3.62x**
StringWalk_utf16_characters_Backwards      | 27240  | 10240 | -62.4% | **2.66x**
SuffixCountableRange                       | 13     | 6     | -53.8% | **2.17x**
StringHasSuffixUnicode                     | 100000 | 48000 | -52.0% | **2.08x (?)**
Data.append.Sequence.64kB.Count.RE.I       | 2      | 1     | -50.0% | **2.00x (?)**
StringHasPrefixUnicode                     | 60000  | 32000 | -46.7% | **1.87x (?)**
CSVParsing.Char                            | 151    | 94    | -37.7% | **1.61x**
StringUTF16SubstringBuilder                | 1620   | 1090  | -32.7% | **1.49x (?)**
CharacterPropertiesStashed                 | 550    | 380   | -30.9% | **1.45x (?)**
CharacterPropertiesPrecomputed             | 690    | 520   | -24.6% | **1.33x**
LineSink.characters.alpha                  | 52     | 40    | -23.1% | **1.30x**
ArrayAppendOptionals                       | 920    | 710   | -22.8% | **1.30x (?)**
StringWalk_tweet_characters_Backwards      | 4920   | 4040  | -17.9% | **1.22x (?)**
CharacterPropertiesStashedMemo             | 800    | 660   | -17.5% | **1.21x (?)**
LineSink.characters.complex                | 277    | 230   | -17.0% | **1.20x (?)**
DataCreateMedium                           | 1200   | 1000  | -16.7% | **1.20x (?)**
CharacterPropertiesFetch                   | 1510   | 1350  | -10.6% | **1.12x (?)**
StringWalk_tweet_characters                | 3320   | 3000  | -9.6%  | **1.11x (?)**
NSArray.bridged.objectAtIndex              | 229    | 208   | -9.2%  | **1.10x (?)**
ParseInt.UInt64.Hex                        | 243    | 222   | -8.6%  | **1.09x (?)**

### Code size: -O


### Performance (arm64): -Osize

**Regression**                                     | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                                               | ---:   | ---:  | ---:    | ---:     
SuffixSequenceLazy                                 | 117    | 828   | +607.7% | **0.14x**
SuffixAnySequence                                  | 117    | 802   | +585.5% | **0.15x**
SuffixSequence                                     | 117    | 802   | +585.5% | **0.15x**
DataCreateEmptyArray                               | 800    | 950   | +18.7%  | **0.84x (?)**
DataToStringEmpty                                  | 400    | 450   | +12.5%  | **0.89x (?)**
SuffixAnySequenceLazy                              | 1633   | 1814  | +11.1%  | **0.90x (?)**
DataCreateSmallArray                               | 1650   | 1800  | +9.1%   | **0.92x (?)**
StringWalk_ascii_characters_Backwards              | 1880   | 2040  | +8.5%   | **0.92x (?)**
StringWalk_punctuated_characters_Backwards         | 480    | 520   | +8.3%   | **0.92x (?)**
StringWalk_punctuatedJapanese_characters_Backwards | 480    | 520   | +8.3%   | **0.92x (?)**
SortArrayInClass                                   | 58036  | 62725 | +8.1%   | **0.93x (?)**
&nbsp; | | | | 
**Improvement**                                    | **OLD** | **NEW** | **DELTA** | **RATIO**
SubstringTrimmingASCIIWhitespace                   | 317    | 88    | -72.2%  | **3.60x**
StringWalk_utf16_characters                        | 26200  | 7280  | -72.2%  | **3.60x**
StringWalk_utf16_characters_Backwards              | 27160  | 10280 | -62.2%  | **2.64x**
StringHasSuffixUnicode                             | 101000 | 47000 | -53.5%  | **2.15x (?)**
StringHasPrefixUnicode                             | 60000  | 32000 | -46.7%  | **1.87x (?)**
CSVParsing.Char                                    | 158    | 101   | -36.1%  | **1.56x**
StringUTF16SubstringBuilder                        | 1670   | 1130  | -32.3%  | **1.48x (?)**
CharacterPropertiesStashedMemo                     | 840    | 580   | -31.0%  | **1.45x (?)**
CharacterPropertiesStashed                         | 630    | 440   | -30.2%  | **1.43x (?)**
SIMDReduce.Int8x16.Cast                            | 43     | 31    | -27.9%  | **1.39x (?)**
CharacterPropertiesPrecomputed                     | 680    | 500   | -26.5%  | **1.36x**
LineSink.characters.alpha                          | 57     | 45    | -21.1%  | **1.27x**
StringWalk_tweet_characters_Backwards              | 5120   | 4240  | -17.2%  | **1.21x (?)**
LineSink.characters.complex                        | 294    | 247   | -16.0%  | **1.19x (?)**
CharacterPropertiesFetch                           | 1550   | 1380  | -11.0%  | **1.12x (?)**
StringWalk_tweet_characters                        | 3560   | 3240  | -9.0%   | **1.10x (?)**
ConvertFloatingPoint.MockFloat64ToInt64            | 433    | 402   | -7.2%   | **1.08x (?)**
FindString.Rec3.Substring                          | 75     | 70    | -6.7%   | **1.07x (?)**

### Code size: -Osize


### Performance (arm64): -Onone

**Regression**                        | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                                  | ---:   | ---:  | ---:   | ---:     
Breadcrumbs.MutatedUTF16ToIdx.ASCII   | 2      | 3     | +50.0% | **0.67x (?)**
ArrayOfGenericPOD2                    | 902    | 1029  | +14.1% | **0.88x (?)**
String.replaceSubrange.String         | 12     | 13    | +8.3%  | **0.92x (?)**
RawBufferCopyBytes                    | 12     | 13    | +8.3%  | **0.92x (?)**
&nbsp; | | | | 
**Improvement**                       | **OLD** | **NEW** | **DELTA** | **RATIO**
StringWalk_utf16_characters           | 29720  | 10360 | -65.1% | **2.87x**
StringWalk_utf16_characters_Backwards | 31560  | 13960 | -55.8% | **2.26x**
StringHasSuffixUnicode                | 102000 | 49000 | -52.0% | **2.08x (?)**
StringHasPrefixUnicode                | 61000  | 33000 | -45.9% | **1.85x (?)**
SubstringTrimmingASCIIWhitespace      | 648    | 401   | -38.1% | **1.62x (?)**
ArrayPlusEqualArrayOfInt              | 220    | 170   | -22.7% | **1.29x (?)**
CharacterPropertiesStashed            | 1010   | 830   | -17.8% | **1.22x (?)**
CSVParsing.Char                       | 334    | 275   | -17.7% | **1.21x (?)**
StringUTF16SubstringBuilder           | 4390   | 3710  | -15.5% | **1.18x (?)**
LineSink.characters.alpha             | 71     | 61    | -14.1% | **1.16x (?)**
CharacterPropertiesPrecomputed        | 1430   | 1230  | -14.0% | **1.16x (?)**
LineSink.characters.complex           | 330    | 285   | -13.6% | **1.16x (?)**
CharacterPropertiesStashedMemo        | 1700   | 1500  | -11.8% | **1.13x (?)**
CharacterPropertiesFetch              | 1940   | 1760  | -9.3%  | **1.10x (?)**
StringWalk_tweet_characters_Backwards | 13680  | 12680 | -7.3%  | **1.08x (?)**
DataReplaceSmall                      | 1500   | 1400  | -6.7%  | **1.07x (?)**

### Code size: -swiftlibs



<details>
  <summary><strong>How to read the data</strong></summary>
The tables contain differences in performance which are larger than 8% and
differences in code size which are larger than 1%.

If you see any unexpected regressions, you should consider fixing the
regressions before you merge the PR.

**Noise**: Sometimes the performance results (not code size!) contain false
alarms. Unexpected regressions which are marked with '(?)' are probably noise.
If you see regressions which you cannot explain you can try to run the
benchmarks again. If regressions still show up, please consult with the
performance team (@eeckstein).
</details>

<details>
  <summary><strong>Hardware Overview</strong></summary>

      Model Name: MacBook Pro
      Model Identifier: MacBookPro17,1
      Total Number of Cores: 8 (4 performance and 4 efficiency)
      Memory: 16 GB

</details>

**Pull request merged**
